canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+                Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                cwRequest.SetNamespace("KinesisVideoSDKCanaryCPP");
+
+                auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                auto persistedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatency_datum, data->pDimension_per_stream, cwRequest);
+                LOG_DEBUG("Persisted Ack Latency: " << persistedAckLatency);
+                if (data->pCanaryConfig->useAggMetrics)
+                {
+                    pushMetric("PersistedAckLatency", persistedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, persistedAckLatency_datum, data->pAggregated_dimension, cwRequest);
+
+                }
+                auto outcome = data->pCWclient->PutMetricData(cwRequest);
+                if (!outcome.IsSuccess())
+                {
+                    cout << "Failed to put PersistedAckLatency metric data:" << outcome.GetError().GetMessage() << endl;
+                }
+                else
+                {
+                    cout << "Successfully put PersistedAckLatency metric data" << endl;
+                }
+            } else if (pFragmentAck->ackType == FRAGMENT_ACK_TYPE_RECEIVED)
+                {
+                    Aws::CloudWatch::Model::MetricDatum receivedAckLatency_datum;
+                    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+                    cwRequest.SetNamespace("KinesisVideoSDKCanaryCPP");
+
+                    auto currentTimestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+                    auto receivedAckLatency = (currentTimestamp - timeOfFragmentEndSent); // [milliseconds]
+                    pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pDimension_per_stream, cwRequest);
+                    LOG_DEBUG("Received Ack Latencyy: " << receivedAckLatency);
+                    if (data->pCanaryConfig->useAggMetrics)
+                    {
+                        pushMetric("ReceivedAckLatency", receivedAckLatency, Aws::CloudWatch::Model::StandardUnit::Milliseconds, receivedAckLatency_datum, data->pAggregated_dimension, cwRequest);
+                    }
+
+                    auto outcome = data->pCWclient->PutMetricData(cwRequest);
+                    if (!outcome.IsSuccess())
+                    {
+                        cout << "Failed to put ReceivedAckLatency metric data:" << outcome.GetError().GetMessage() << endl;
+                    }
+                    else
+                    {
+                        cout << "Successfully put ReceivedAckLatency metric data" << endl;
+                    }
+                }
+        } else
```

Comment:
This else portion does not make sense. I think what you are going with if it neither received or persisted ack. However, we still want to track other acks (error and buffered( even if we are not going to publish metrics for them. It can be a simple log line, but we need it. So when you modify this to switch case, let us add error ack and buffered ack type as well.